### ISSUE TITLE
StubbedRequests fails when body contains a listed form file

### ISSUE THREAD

**Issue by aaronwolen** [ropensci/webmockr#84]
Ran into this while finishing up on https://github.com/ropensci/vcr/pull/130. I initially thought it was an issue caused by the new `pluck_body()` but I can reproduce it on the master branch w/ either crul or httr.

``` r
# remotes::install_github("ropensci/webmockr")
library(webmockr)

cli <- crul::HttpClient$new("https://httpbin.org")
body <- list(y = crul::upload(system.file("CITATION")))

enable()
#> CrulAdapter enabled!
#> HttrAdapter enabled!
b <- cli$post("post", body = body)
#> Error in paste(names(x)[i], switch(class(targ)[1L], character = sprintf("\"%s\"", : non-string argument to internal 'paste'
httr::POST("https://httpbin.org/post", body = body)
#> Error in paste(names(x)[i], switch(class(targ)[1L], character = sprintf("\"%s\"", : non-string argument to internal 'paste'
```

I tracked it back to `RequestSignature`'s `to_s()`method but didn't get to dig much further. 

```r
devtools::session_info()
#> Warning in system("timedatectl", intern = TRUE): running command 'timedatectl'
#> had status 1
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 3.6.0 (2019-04-26)
#>  os       Ubuntu 16.04.6 LTS          
#>  system   x86_64, linux-gnu           
#>  ui       X11                         
#>  language (EN)                        
#>  collate  C.UTF-8                     
#>  ctype    C.UTF-8                     
#>  tz       Etc/UTC                     
#>  date     2019-12-17                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version  date       lib source                            
#>  assertthat    0.2.1    2019-03-21 [1] RSPM (R 3.6.0)                    
#>  backports     1.1.5    2019-10-02 [1] RSPM (R 3.6.0)                    
#>  callr         3.4.0    2019-12-09 [1] RSPM (R 3.6.0)                    
#>  cli           2.0.0    2019-12-09 [1] RSPM (R 3.6.0)                    
#>  crayon        1.3.4    2017-09-16 [1] RSPM (R 3.6.0)                    
#>  crul          0.9.0    2019-11-06 [1] RSPM (R 3.6.0)                    
#>  curl          4.3      2019-12-02 [1] RSPM (R 3.6.0)                    
#>  desc          1.2.0    2018-05-01 [1] RSPM (R 3.6.0)                    
#>  devtools      2.2.1    2019-09-24 [1] RSPM (R 3.6.0)                    
#>  digest        0.6.23   2019-11-23 [1] RSPM (R 3.6.0)                    
#>  ellipsis      0.3.0    2019-09-20 [1] RSPM (R 3.6.0)                    
#>  evaluate      0.14     2019-05-28 [1] RSPM (R 3.6.0)                    
#>  fansi         0.4.0    2018-10-05 [1] RSPM (R 3.6.0)                    
#>  fauxpas       0.2.0    2018-03-01 [1] RSPM (R 3.6.0)                    
#>  fs            1.3.1    2019-05-06 [1] RSPM (R 3.6.0)                    
#>  glue          1.3.1    2019-03-12 [1] RSPM (R 3.6.0)                    
#>  highr         0.8      2019-03-20 [1] RSPM (R 3.6.0)                    
#>  htmltools     0.4.0    2019-10-04 [1] RSPM (R 3.6.0)                    
#>  httpcode      0.2.0    2016-11-14 [1] RSPM (R 3.6.0)                    
#>  httr          1.4.1    2019-08-05 [1] RSPM (R 3.6.0)                    
#>  knitr         1.26     2019-11-12 [1] RSPM (R 3.6.0)                    
#>  magrittr      1.5      2014-11-22 [1] RSPM (R 3.6.0)                    
#>  memoise       1.1.0    2017-04-21 [1] RSPM (R 3.6.0)                    
#>  mime          0.7      2019-06-11 [1] RSPM (R 3.6.0)                    
#>  pkgbuild      1.0.6    2019-10-09 [1] RSPM (R 3.6.0)                    
#>  pkgload       1.0.2    2018-10-29 [1] RSPM (R 3.6.0)                    
#>  prettyunits   1.0.2    2015-07-13 [1] RSPM (R 3.6.0)                    
#>  processx      3.4.1    2019-07-18 [1] RSPM (R 3.6.0)                    
#>  ps            1.3.0    2018-12-21 [1] RSPM (R 3.6.0)                    
#>  R6            2.4.1    2019-11-12 [1] RSPM (R 3.6.0)                    
#>  Rcpp          1.0.3    2019-11-08 [1] RSPM (R 3.6.0)                    
#>  remotes       2.1.0    2019-06-24 [1] RSPM (R 3.6.0)                    
#>  rlang         0.4.2    2019-11-23 [1] RSPM (R 3.6.0)                    
#>  rmarkdown     2.0      2019-12-12 [1] RSPM (R 3.6.0)                    
#>  rprojroot     1.3-2    2018-01-03 [1] RSPM (R 3.6.0)                    
#>  sessioninfo   1.1.1    2018-11-05 [1] RSPM (R 3.6.0)                    
#>  stringi       1.4.3    2019-03-12 [1] RSPM (R 3.6.0)                    
#>  stringr       1.4.0    2019-02-10 [1] RSPM (R 3.6.0)                    
#>  testthat      2.3.1    2019-12-01 [1] RSPM (R 3.6.0)                    
#>  triebeard     0.3.0    2016-08-04 [1] RSPM (R 3.6.0)                    
#>  urltools      1.7.3    2019-04-14 [1] RSPM (R 3.6.0)                    
#>  usethis       1.5.1    2019-07-04 [1] RSPM (R 3.6.0)                    
#>  webmockr    * 0.5.1.91 2019-12-17 [1] Github (ropensci/webmockr@8e6fd9a)
#>  withr         2.1.2    2018-03-15 [1] RSPM (R 3.6.0)                    
#>  xfun          0.11     2019-11-12 [1] RSPM (R 3.6.0)                    
#>  yaml          2.2.0    2018-07-25 [1] RSPM (R 3.6.0)                    
#> 
#> [1] /home/rstudio-user/R/x86_64-pc-linux-gnu-library/3.6
#> [2] /opt/R/3.6.0/lib/R/library
```

<sup>Created on 2019-12-17 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by sckott** [comment]
thanks! will have a look

**Comment by sckott** [comment]
working on this 😓 its complicated ... I'm probably going to have to split apart `pluck_body` - a couple different scenarios

- when used for constructing a real httr request https://github.com/ropensci/webmockr/blob/master/R/adapter-httr.R#L167-L173 when net connect is allowed when webmockr is enabled, as the body that we need there what you'd actually pass to body in httr verb, for an upload not in a list 
- `pluck_body` for webmockr in some cases is used just as a representation of the body we can use to summarise the request or similar
- in vcr - `pluck_body` doesn't seem to affect caching requests and replaying them, but it does affect what gets put in the cassette request slot - & therefore what gets put in the `$request` slot in the "fake" responses re-created from cassettes

**Comment by aaronwolen** [comment]
It *is* complicated. I enjoyed poking around the internals of webmockr/vcr and learning how they work but I had to sketch out a diagram of the various classes and how they interact with one another. It looked like one of those bulletin boards from Homeland.

Does it make sense to add something like `format_body()` (or `print_body()`) for the express purpose of formatting the various body types? That might help make it clear when the original body object(s) is needed vs when we're just returning a representation of it.

**Comment by sckott** [comment]
Good idea to map it out. Would be good to add something like that to the book https://books.ropensci.org/http-testing/

Perhaps yes on breaking out fxns for making string representations of bodies vs. the actual body - hopefully will push something soon

**Comment by sckott** [comment]
fix in 63c9760
